data/transactions/logic/box.go

```diff
@@ -207,28 +231,31 @@ func opBoxPut(cx *EvalContext) error {
 	value := string(cx.stack[last].Bytes)
 	name := string(cx.stack[prev].Bytes)
 
-	err := cx.availableBox(name, boxWrite, 0 /* unused for write */)
-	if err != nil {
-		return err
-	}
-	box, ok, err := cx.Ledger.GetBox(cx.appID, name)
+	contents, exists, err := cx.availableBox(name, boxWrite, uint64(len(value)))
 	if err != nil {
 		return err
 	}
 
 	cx.stack = cx.stack[:prev]
 
-	if ok {
-		/* the box exists, so the replacement must match size */
-		if len(box) != len(value) {
-			return fmt.Errorf("attempt to box_put wrong size %d != %d", len(box), len(value))
+	if exists {
+		/* the replacement must match existing size */
+		if len(contents) != len(value) {
+			return fmt.Errorf("attempt to box_put wrong size %d != %d", len(contents), len(value))
 		}
 		return cx.Ledger.SetBox(cx.appID, name, value)
 	}
+
 	/* The box did not exist, so create it. */
+	_, err = createCheck(cx, name, uint64(len(contents)))
+	if err != nil {
+		return err
+	}
 	appAddr := cx.getApplicationAddress(cx.appID)
-
-	_, err = createBox(cx, name, value, appAddr)
+	created, err := cx.Ledger.NewBox(cx.appID, name, value, appAddr)
+	if !created {
+		panic(name)
+	}
 	return err
```

Comment:
It's certainly should never trigger. I don't feel strongly it should remain (I think I intended to remove it).

I'm never sure whether we should leave stuff like this in.

Reply:
Even better - since we now decide whether to call NewBox based on whether we _need_ to create or not, I can revert that interface to failing if it can't create.

Reply:
done

Reply:
@jannotti Thanks - I had been pondering the interface change, but wasn't sure if there was a good reason to retain it.  Change looks good.